src/vstest.console/TestPlatformHelpers/TestRequestManager.cs

```diff
@@ -142,7 +157,10 @@ public bool DiscoverTests(DiscoveryRequestPayload discoveryPayload, ITestDiscove
             var requestData = new RequestData
                                   {
                                       ProtocolConfig = protocolConfig,
-                                      MetricsCollection = new MetricsCollection()
+                                      MetricsCollection =
+                                          this.telemetryOptedOut
+                                              ? (IMetricsCollection)new NoOpMetricsCollection()
+                                              : new MetricsCollection()
```

Comment:
nit : Add this as another property which decides based on the optedout or in. So you don't need to duplicate the code and looks clean.